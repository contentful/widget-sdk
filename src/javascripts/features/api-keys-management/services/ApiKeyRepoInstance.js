import { getModule } from 'core/NgRegistry';
import { createApiKeyRepo } from './ApiKeyRepo';

let apiKeyRepoInstance = null;

export function getApiKeyRepo() {
  const spaceContext = getModule('spaceContext');

  if (!apiKeyRepoInstance) {
    apiKeyRepoInstance = createApiKeyRepo(spaceContext.endpoint);
  }

  return apiKeyRepoInstance;
}

export function purgeApiKeyRepoCache() {
  apiKeyRepoInstance = null;
}
