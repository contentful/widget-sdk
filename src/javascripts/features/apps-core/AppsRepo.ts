import { omit, sortBy } from 'lodash';

import { fetchMarketplaceApps, fetchContentfulApps } from './MarketplaceClient';
import type APIClient from 'data/APIClient';
import { AppDefinition, AppInstallation } from 'contentful-management/types';
import { MarketplaceApp } from './types';

interface AppsRepoCache {
  marketplaceAppDefinitions?: Promise<MarketplaceApp[]>;
  contentfulAppDefinitions?: Promise<MarketplaceApp[]>;
  appInstallations?: Promise<any>;
}

function makePrivateAppObject(
  appDefinition: AppDefinition,
  appInstallation: AppInstallation
): MarketplaceApp {
  return {
    id: appDefinition.sys.id,
    title: appDefinition.name,
    appDefinition,
    appInstallation,
    isPrivateApp: true,
    isEarlyAccess: false, // private apps can never be early access
  };
}

function getPrivateApps(
  installationMap: Record<string, AppInstallation>,
  orgDefinitions
): MarketplaceApp[] {
  return orgDefinitions.map((def) => {
    return makePrivateAppObject(def, installationMap[def.sys.id]);
  });
}

function makeMarketplaceAppObject(
  marketplaceApp: MarketplaceApp,
  appDefinition: AppDefinition,
  appInstallation: AppInstallation
) {
  return {
    ...omit(marketplaceApp, ['definitionId']),
    appDefinition,
    appInstallation,
  };
}

export class AppsRepo {
  cache: AppsRepoCache = {};
  _cacheKey = {};

  private cacheInvalidatedListeners: Set<() => void> = new Set();

  constructor(private cma: APIClient, private appDefinitionLoader) {}

  onCacheInvalidated(listener: () => unknown) {
    const uniqueListener = () => listener();

    // this allows hooks to add a listener to trigger a rerendewr
    this.cacheInvalidatedListeners.add(uniqueListener);
    return () => this.cacheInvalidatedListeners.delete(uniqueListener);
  }

  clearCache(fields = Object.keys(this.cache) as (keyof AppsRepoCache)[]) {
    fields.forEach((field) => delete this.cache[field]);
    // this is needed for hooks to compare invalidate memos when they render
    this._cacheKey = {};
    this.cacheInvalidatedListeners.forEach((callback) => callback());
  }

  getAppByIdOrSlug = async (idOrSlug: string): Promise<MarketplaceApp> => {
    const apps = await this.getAllApps();

    const app = apps.find((app) => app.id === idOrSlug || app.appDefinition.sys.id === idOrSlug);
    if (app) {
      return app;
    }

    throw new Error(`Could not find an app with ID "${idOrSlug}".`);
  };

  getAllApps = async (): Promise<MarketplaceApp[]> => {
    const orgDefinitions = await this.appDefinitionLoader.getAllForCurrentOrganization();

    const [installationMap, resolvedMarketplaceApps, resolvedContentfulApps] = await Promise.all([
      this.getAppDefinitionToInstallationMap(),
      this.getMarketplaceApps(),
      this.getContentfulApps(),
    ]);

    return [
      ...resolvedContentfulApps,
      ...resolvedMarketplaceApps,
      ...getPrivateApps(installationMap, orgDefinitions),
    ];
  };

  async getContentfulApps(): Promise<MarketplaceApp[]> {
    const [installationMap, fetchedApps] = await Promise.all([
      this.getAppDefinitionToInstallationMap(),
      this.getContentfulAppDefinitions(),
    ]);
    const ctfAppDefs = await this.appDefinitionLoader.getByIds(
      fetchedApps?.map((c) => c.definitionId)
    );

    return sortBy(
      fetchedApps
        ?.map((app) => ({
          ...app,
          appDefinition: ctfAppDefs[app.definitionId!],
          appInstallation: installationMap[app.definitionId!],
        }))
        .filter((app) => app.appDefinition),
      (ctflApp) => ctflApp.title.toLowerCase()
    );
  }

  getOnlyInstalledApps = async (): Promise<MarketplaceApp[]> => {
    const [appInstallations, marketplaceApps, contentfulApps] = await Promise.all([
      this.getAppInstallations(),
      this.getMarketplaceAppDefinitions(),
      this.getContentfulAppDefinitions(),
    ]);
    const availableApps = marketplaceApps.concat(contentfulApps);

    return appInstallations.items.map((appInstallation) => {
      const definitionId = appInstallation.sys.appDefinition.sys.id;
      const appDefinition = appInstallations.includes.AppDefinition.find((def) => {
        return def.sys.id === definitionId;
      });
      const marketplaceApp = availableApps.find((app) => {
        return app.definitionId === definitionId;
      });

      if (marketplaceApp) {
        return makeMarketplaceAppObject(marketplaceApp, appDefinition, appInstallation);
      } else {
        return makePrivateAppObject(appDefinition, appInstallation);
      }
    });
  };

  private getCachedOrCreate<Key extends keyof AppsRepoCache>(
    key: Key,
    getNewValue: () => AppsRepoCache[Key]
  ) {
    const cachedValue = this.cache[key];

    if (cachedValue) return cachedValue!;

    const newValue = getNewValue();
    this.cache[key] = newValue;

    return newValue!;
  }

  private async getMarketplaceAppDefinitions() {
    return this.getCachedOrCreate('marketplaceAppDefinitions', fetchMarketplaceApps);
  }

  private async getContentfulAppDefinitions() {
    return this.getCachedOrCreate('contentfulAppDefinitions', fetchContentfulApps);
  }

  private async getAppInstallations() {
    return this.getCachedOrCreate('appInstallations', () => this.cma.getAppInstallations());
  }

  private async getMarketplaceApps(): Promise<MarketplaceApp[]> {
    const [installationMap, fetchedApps] = await Promise.all([
      this.getAppDefinitionToInstallationMap(),
      this.getMarketplaceAppDefinitions(),
    ]);
    const definitionIds = fetchedApps
      // show apps from the marketplace space which are in the public listing or are installed
      .filter((app) => app.isListed || !!installationMap[app.definitionId!])
      .map((app) => app.definitionId);
    const appDefinitionMap = await this.appDefinitionLoader.getByIds(definitionIds);

    return sortBy(
      fetchedApps.reduce((acc, app) => {
        const definitionId = app.definitionId!;
        const appDefinition = appDefinitionMap[definitionId];
        const appInstallation = installationMap[definitionId];

        // Referred definition, for whatever reason, is missing. Skip the app.
        if (!appDefinition) {
          return acc;
        }

        return [...acc, makeMarketplaceAppObject(app, appDefinition, appInstallation)];
      }, [] as MarketplaceApp[]),
      (app) => app.title.toLowerCase()
    );
  }

  private async getAppDefinitionToInstallationMap() {
    const { items } = await this.getAppInstallations();

    return items.reduce(
      (acc, installation) => ({
        ...acc,
        [installation.sys.appDefinition.sys.id]: installation,
      }),
      {}
    );
  }
}
