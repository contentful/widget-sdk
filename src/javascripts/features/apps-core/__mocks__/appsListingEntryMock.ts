export const appsListingEntryMock = {
  sys: {
    type: 'Array',
  },
  total: 1,
  skip: 0,
  limit: 100,
  items: [
    {
      sys: {
        space: {
          sys: {
            type: 'Link',
            linkType: 'Space',
            id: 'lpjm8d10rkpy',
          },
        },
        id: '2fPbSMx3baxlwZoCyXC7F1',
        type: 'Entry',
        createdAt: '2019-10-02T14:57:50.083Z',
        updatedAt: '2020-02-24T13:00:26.435Z',
        environment: {
          sys: {
            id: 'master',
            type: 'Link',
            linkType: 'Environment',
          },
        },
        revision: 30,
        contentType: {
          sys: {
            type: 'Link',
            linkType: 'ContentType',
            id: 'appsListing',
          },
        },
        locale: 'en-US',
      },
      fields: {
        title: 'Apps Listing',
        apps: [
          {
            sys: {
              type: 'Link',
              linkType: 'Entry',
              id: '5ZSCLEfnoPdTF3agzaGGRb',
            },
          },
          {
            sys: {
              type: 'Link',
              linkType: 'Entry',
              id: '4iIetqcwsR1GIZxaYI6fRm',
            },
          },
        ],
      },
    },
  ],
};
