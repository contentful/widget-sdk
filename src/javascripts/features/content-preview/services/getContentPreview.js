import { getSpaceContext } from 'classes/spaceContext';
import { createContentPreview } from './createContentPreview';

let contentPreviewInstance = null;

export function getContentPreview() {
  const spaceContext = getSpaceContext();
  const space = spaceContext.getSpace();

  if (!contentPreviewInstance) {
    contentPreviewInstance = createContentPreview({ space, cma: spaceContext.cma });
  }

  return contentPreviewInstance;
}

export function purgeContentPreviewCache() {
  contentPreviewInstance = null;
}
