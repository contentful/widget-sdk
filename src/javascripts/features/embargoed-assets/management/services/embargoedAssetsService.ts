import * as EndpointFactory from 'data/EndpointFactory';
import APIClient from 'data/APIClient';
import { LEVEL } from '../constants';

export interface EmbargoedAsset {
  level: LEVEL;
}

export function embargoedAssets(spaceId: string | undefined) {
  const endpoint = EndpointFactory.createSpaceEndpoint(spaceId, undefined);
  const apiClient = new APIClient(endpoint);

  return {
    getCurrentLevel() {
      return apiClient.getEmbargoedAssetsSettingLevel() as Promise<EmbargoedAsset>;
    },
    setCurrentLevel(level: LEVEL | 'enabled') {
      return apiClient.setEmbargoedAssetsSettingLevel(level) as Promise<EmbargoedAsset>;
    },
  };
}
