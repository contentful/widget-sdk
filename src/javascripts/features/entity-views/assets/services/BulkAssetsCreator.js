import { identity, isString } from 'lodash';
import { FilestackService } from '@contentful/field-editor-file';
import * as Config from 'Config';
import { Notification } from '@contentful/forma-36-react-components';
import * as stringUtils from 'utils/StringUtils';
import * as logger from 'services/logger';
import { getModule } from 'core/NgRegistry';
import { getCMAClient } from 'core/services/usePlainCMAClient';

/**
 * @typedef { import("contentful-management").PlainClientAPI }  PlainClientAPI
 */

/**
 * Opens file selector to select files which will then be uploaded as assets.
 * The resolved assets will still be in processing once `open()` resolves. Any
 * remote action triggered on these instances will fail as they are invalid
 * as long as processing is not done and once it is done the version will be
 * outdated.
 *
 * @param {string} localeCode Internal locale code files will be uploaded for.
 * @returns {Promise<Array<Asset>>}
 */
export function open(localeCode) {
  if (!isString(localeCode)) {
    throw new TypeError('locale must be a string');
  }

  const spaceContext = getModule('spaceContext');

  const cma = getCMAClient({
    spaceId: spaceContext.getId(),
    environmentId: spaceContext.getEnvironmentId(),
  });

  const { apiKey, policy, signature } = Config.services.filestack;

  return FilestackService.pickMultiple({
    config: {
      apiKey,
      policy,
      signature,
    },
  }).then(createAssetsForFiles, () => {
    Notification.error(
      'An error occurred while uploading multiple assets. ' +
        'Please contact support if this problem persists.'
    );

    return [];
  });

  function createAssetsForFiles(files) {
    if (files.length === 0) {
      return Promise.resolve([]);
    }
    return Promise.all(files.map(createAssetForFile)).then(
      (assets) => {
        assets = assets.filter(identity);
        Notification.success('Assets uploaded. Processing…');
        return Promise.all(assets.map(processAssetForFile))
          .then(() => {
            Notification.success('Assets processed. Updating…');
            return assets;
          })
          .catch((error) => {
            Notification.error('Some assets failed to process');
            return Promise.reject(error);
          });
      },
      (error) => {
        logger.captureWarning(error);
        Notification.error('Some assets failed to upload');

        return Promise.reject(error);
      }
    );
  }

  function createAssetForFile(file) {
    const title = stringUtils.fileNameToTitle(file.fileName);
    return cma.asset.create(
      {},
      {
        fields: {
          file: {
            [localeCode]: file,
          },
          title: { [localeCode]: title },
        },
      }
    );
  }

  function processAssetForFile(asset) {
    return cma.asset.processForLocale({}, asset, localeCode);
  }
}
