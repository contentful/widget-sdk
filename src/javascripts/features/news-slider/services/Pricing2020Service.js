import React from 'react';
import { getModule } from 'core/NgRegistry';
import { getOrganization, getSpace } from 'services/TokenStore';
import { getSubscriptionPlans } from 'account/pricing/PricingDataProvider';
import { createOrganizationEndpoint } from 'data/EndpointFactory';
import { isOwnerOrAdmin } from 'services/OrganizationRoles';
import { getAllSpaces } from 'access_control/OrganizationMembershipRepository';
import { getVariation } from 'LaunchDarkly';
import { PRICING_2020_WARNING } from 'featureFlags';
import { SelfService } from '../components/Pricing2020';
import { ModalLauncher } from 'core/components/ModalLauncher';
import { createClientStorage } from 'core/services/BrowserStorage/ClientStorage';

// TODO: move this somewhere else
export async function getCurrentOrg() {
  const { orgId, spaceId } = getModule('$stateParams');

  if (orgId) {
    return getOrganization(orgId);
  } else if (spaceId) {
    const space = await getSpace(spaceId);
    return space.sys.organization;
  }

  return null;
}

export async function openPricing2020Warning() {
  const org = await getCurrentOrg();

  if (!(await shouldDisplay(org))) {
    return;
  }

  const endpoint = createOrganizationEndpoint(org.sys.id);

  const { items: plans } = await getSubscriptionPlans(endpoint);
  const spaces = await getAllSpaces(endpoint);

  const basePlan = plans.find((plan) => plan.planType === 'base');
  const spacePlans = plans.filter((plan) => plan.planType === 'space');
  // TODO: filter by search plan id instead
  const microPlans = spacePlans.filter((plan) => plan.name === 'Micro');
  const microSpaces = microPlans.map(
    (plan) => spaces.find((space) => space.sys.id === plan.gatekeeperKey)?.name
  );
  const freeSpace = spaces.find(
    (space) => !spacePlans.some((plan) => space.sys.id === plan.gatekeeperKey)
  )?.name;
  const isCommunity = basePlan.name === 'Community Platform';
  const isTeam = basePlan.name === 'Team';

  // TODO: return early if free space was created after the migration

  // if org has no free spaces, there's nothing to see
  if (!freeSpace) return;

  if (isCommunity || isTeam) {
    const key = getStorageKey(org.sys.id);
    const storage = createClientStorage('local');
    storage.set(key, 'true');
  }

  if (isCommunity) {
    // TODO: display the community slides when they're ready
    ModalLauncher.open(({ isShown, onClose }) => (
      <SelfService
        isShown={isShown}
        onClose={onClose}
        freeSpace={freeSpace}
        microSpaces={microSpaces}
        orgId={org.sys.id}
      />
    ));
  } else if (isTeam) {
    ModalLauncher.open(({ isShown, onClose }) => (
      <SelfService
        isShown={isShown}
        onClose={onClose}
        freeSpace={freeSpace}
        microSpaces={microSpaces}
        orgId={org.sys.id}
      />
    ));
  }
}

async function shouldDisplay(org) {
  if (!org) return false;
  if (hasSeen(org.sys.id)) return false;

  const isEnabled = await getVariation(PRICING_2020_WARNING, { organizationId: org.sys.id });
  const isTargetUser = isOwnerOrAdmin(org);

  return isEnabled && isTargetUser;
}

function hasSeen(orgId) {
  const key = getStorageKey(orgId);
  const storage = createClientStorage('local');
  return storage.get(key);
}

function getStorageKey(orgId) {
  const storagePrefix = 'hasSeenPricing2020WarningForOrg';
  return `${storagePrefix}.${orgId}`;
}
