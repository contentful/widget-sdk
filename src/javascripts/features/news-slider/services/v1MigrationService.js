import React from 'react';
import { createOrganizationEndpoint } from 'data/EndpointFactory';
import { isOwnerOrAdmin } from 'services/OrganizationRoles';
import { getVariation, FLAGS } from 'core/feature-flags';
import { GenericCustomerSlides } from '../components';
import { ModalLauncher } from '@contentful/forma-36-react-components/dist/alpha';
import { createClientStorage } from 'core/services/BrowserStorage/ClientStorage';
import { track } from 'analytics/Analytics';
import { getModule } from 'core/NgRegistry';
import { getOrganization, getSpace } from 'services/TokenStore';
import { getBasePlan } from 'features/pricing-entities';
import { isFreePlan, isSelfServicePlan } from 'account/pricing/PricingDataProvider';
import { isPartnerPlan, isProBonoPlan } from 'account/pricing/PricingDataProvider';
import { generateBasePlanName } from 'features/organization-subscription';

const V1_DESTINATION_COMMUNITY = 'community';
const V1_DESTINATION_TEAM = 'team';
const V1_DESTINATION_PRO_BONO = 'pro_bono';
const V1_DESTINATION_PARTNER = 'partner';

async function getCurrentOrg() {
  const { orgId, spaceId } = getModule('$stateParams');

  if (orgId) {
    return getOrganization(orgId);
  } else if (spaceId) {
    const space = await getSpace(spaceId);
    return space.sys.organization;
  }

  // if the current page is not in the scope of any organization, return null.
  // i.e. Account settings and Home (for users without org access)
  return null;
}

export async function openV1MigrationWarning() {
  const org = await getCurrentOrg();
  if (!(await shouldDisplay(org))) {
    return;
  }

  const endpoint = createOrganizationEndpoint(org.sys.id);
  const basePlan = await getBasePlan(endpoint);
  const v1migrationDestination = org.sys?._v1Migration?.destination;

  const isCommunity = isFreePlan(basePlan) && v1migrationDestination === V1_DESTINATION_COMMUNITY;
  const isTeam = isSelfServicePlan(basePlan) && v1migrationDestination === V1_DESTINATION_TEAM;
  const isProBono = isProBonoPlan(basePlan) && v1migrationDestination === V1_DESTINATION_PRO_BONO;
  const isPartner = isPartnerPlan(basePlan) && v1migrationDestination === V1_DESTINATION_PARTNER;

  if (isCommunity || isTeam || isProBono || isPartner) {
    const key = getStorageKey(org.sys.id);
    const storage = createClientStorage('local');
    storage.set(key, 'true');
    track('v1_migration_update:communication_seen', { basePlanName: basePlan.name });
  }

  const basePlanName = generateBasePlanName(basePlan, v1migrationDestination);

  if (isCommunity || isTeam || isProBono || isPartner) {
    ModalLauncher.open(({ isShown, onClose }) => (
      <GenericCustomerSlides isShown={isShown} onClose={onClose} basePlanName={basePlanName} />
    ));
  }
}

async function shouldDisplay(org) {
  if (!org) {
    return false;
  }

  const isEnabled = await getVariation(FLAGS.V1_MIGRATION_2021_WARNING, {
    organizationId: org.sys.id,
  });

  if (!isEnabled || hasSeen(org.sys.id)) {
    return false;
  }

  // check for the status of the migration
  const v1migrationSucceeded = org?.sys?._v1Migration?.status === 'succeeded';
  const isTargetUser = isOwnerOrAdmin(org);

  return isTargetUser && v1migrationSucceeded;
}

function hasSeen(orgId) {
  const key = getStorageKey(orgId);
  const storage = createClientStorage('local');

  return storage.get(key);
}

function getStorageKey(orgId) {
  const storagePrefix = 'hasSeenV1MigrationWarning';
  return `${storagePrefix}.${orgId}`;
}
