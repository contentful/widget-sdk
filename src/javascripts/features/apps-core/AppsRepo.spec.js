import { createAppsRepo } from './AppsRepo';
import { window } from 'core/services/window';
import appsListingEntryMock from './__mocks__/appsListingEntryMock.json';
import allAppsMock from './__mocks__/appEntriesMock.json';

const NETLIFY_DEFINITION_ID = '1VchawWvbIClHuMIyxwR5m';

jest.mock('core/services/window', () => ({
  window: {
    fetch: jest.fn(),
    location: {
      search: '',
    },
  },
}));

describe('AppsRepo', () => {
  const netlifyInstallation = {
    sys: {
      type: 'AppInstallation',
      appDefinition: {
        sys: {
          type: 'Link',
          linkType: 'AppDefinition',
          id: NETLIFY_DEFINITION_ID,
        },
      },
    },
  };

  const netlifyDefinition = {
    sys: { type: 'AppDefinition', id: NETLIFY_DEFINITION_ID },
    name: 'Netlify',
    locations: ['app', 'entry-sidebar'],
    src: 'http://localhost:1234',
  };

  const privateDefinition = {
    sys: { type: 'AppDefinition', id: 'app-definition-id' },
    name: 'My app',
    locations: ['app', 'entry-field'],
    src: 'http://localhost:666',
  };

  const loader = {
    getByIds: jest.fn(() => {
      return Promise.resolve({ [NETLIFY_DEFINITION_ID]: netlifyDefinition });
    }),
    getAllForCurrentOrganization: jest.fn(() => {
      return Promise.resolve([privateDefinition]);
    }),
  };

  const cma = {
    getAppInstallations: jest.fn(() => {
      return Promise.resolve({ items: [netlifyInstallation] });
    }),
  };

  describe('getApps', () => {
    it('should only return private apps if the marketplace endpoint returns no data', async () => {
      window.fetch.mockImplementation(() => {
        return Promise.resolve({ ok: false, json: () => Promise.resolve({}) });
      });

      const repo = createAppsRepo(cma, loader);
      const result = await repo.getApps();

      expect(
        window.fetch
      ).toHaveBeenCalledWith(
        'https://cdn.contentful.com/spaces/lpjm8d10rkpy/entries?include=0&sys.id[in]=2fPbSMx3baxlwZoCyXC7F1',
        { headers: { Authorization: 'Bearer XMf7qZNsdNypDfO9TC1NZK2YyitHORa_nIYqYdpnQhk' } }
      );

      expect(
        window.fetch
      ).toHaveBeenCalledWith(
        'https://cdn.contentful.com/spaces/lpjm8d10rkpy/entries?include=10&content_type=app',
        { headers: { Authorization: 'Bearer XMf7qZNsdNypDfO9TC1NZK2YyitHORa_nIYqYdpnQhk' } }
      );

      expect(result).toEqual([
        {
          appDefinition: privateDefinition,
          id: 'app-definition-id',
          title: 'My app',
          isPrivateApp: true,
          isEarlyAccess: false,
        },
      ]);
    });

    it('should only return private apps the marketplace endpoint returns bad data', async () => {
      window.fetch.mockImplementation(() => {
        return Promise.resolve({
          ok: true,
          json: () => Promise.resolve({ items: [], includes: { Assets: [] }, sys: {} }),
        });
      });

      const repo = createAppsRepo(cma, loader);
      const result = await repo.getApps();

      expect(result).toEqual([
        {
          appDefinition: privateDefinition,
          id: 'app-definition-id',
          title: 'My app',
          isPrivateApp: true,
          isEarlyAccess: false,
        },
      ]);
    });

    it('should return an array of apps when the marketplace endpoint returns good data', async () => {
      window.fetch.mockImplementation((url) => {
        const jsonResponse = url.endsWith('2fPbSMx3baxlwZoCyXC7F1')
          ? appsListingEntryMock
          : allAppsMock;
        return Promise.resolve({ ok: true, json: () => Promise.resolve(jsonResponse) });
      });

      const repo = createAppsRepo(cma, loader);
      const result = await repo.getApps();

      expect(
        window.fetch
      ).toHaveBeenCalledWith(
        'https://cdn.contentful.com/spaces/lpjm8d10rkpy/entries?include=0&sys.id[in]=2fPbSMx3baxlwZoCyXC7F1',
        { headers: { Authorization: 'Bearer XMf7qZNsdNypDfO9TC1NZK2YyitHORa_nIYqYdpnQhk' } }
      );

      expect(
        window.fetch
      ).toHaveBeenCalledWith(
        'https://cdn.contentful.com/spaces/lpjm8d10rkpy/entries?include=10&content_type=app',
        { headers: { Authorization: 'Bearer XMf7qZNsdNypDfO9TC1NZK2YyitHORa_nIYqYdpnQhk' } }
      );

      expect(result).toMatchSnapshot();
    });
  });

  describe('getApp', () => {
    beforeEach(() => {
      window.fetch.mockImplementation((url) => {
        const jsonResponse = url.endsWith('2fPbSMx3baxlwZoCyXC7F1')
          ? // listing
            appsListingEntryMock
          : // apps
            allAppsMock;

        return Promise.resolve({
          ok: true,
          json: () => Promise.resolve(jsonResponse),
        });
      });
    });

    it('returns an app if present', async () => {
      const repo = createAppsRepo(cma, loader);

      const app1 = await repo.getApp('app-definition-id');
      expect(app1.appDefinition).toEqual(privateDefinition);

      const app2 = await repo.getApp('netlify');
      expect(app2.appDefinition).toEqual(netlifyDefinition);
      expect(app2.appInstallation).toEqual(netlifyInstallation);
    });

    it('fails if an app is not present', async () => {
      const repo = createAppsRepo(cma, loader);

      expect.assertions(1);
      try {
        await repo.getApp('not-here');
      } catch (err) {
        expect(err.message).toBe('Could not find an app with ID "not-here".');
      }
    });

    it('fails if API call fails', async () => {
      const cma = {
        getAppInstallations: () => Promise.reject(new Error('api error')),
      };

      const repo = createAppsRepo(cma, loader);

      expect.assertions(1);
      try {
        await repo.getApp('fails-either-way');
      } catch (err) {
        expect(err.message).toBe('api error');
      }
    });
  });
});
