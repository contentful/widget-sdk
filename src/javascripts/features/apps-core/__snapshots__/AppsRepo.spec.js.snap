// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppsRepo getApps should return an array of apps when the marketplace endpoint returns good data 1`] = `
Array [
  Object {
    "actionList": Array [
      Object {
        "info": "Build widget will be removed from sidebars of all content types",
        "negative": true,
      },
      Object {
        "info": "This will not modify your Netlify sites. You may want to remove unneeded deploy notifications and build hooks from your Netlify site.",
        "negative": false,
      },
    ],
    "appDefinition": Object {
      "locations": Array [
        "app",
        "entry-sidebar",
      ],
      "name": "Netlify",
      "src": "http://localhost:1234",
      "sys": Object {
        "id": "1VchawWvbIClHuMIyxwR5m",
        "type": "AppDefinition",
      },
    },
    "appInstallation": Object {
      "sys": Object {
        "appDefinition": Object {
          "sys": Object {
            "id": "1VchawWvbIClHuMIyxwR5m",
            "linkType": "AppDefinition",
            "type": "Link",
          },
        },
        "type": "AppInstallation",
      },
    },
    "author": Object {
      "icon": "//images.ctfassets.net/lpjm8d10rkpy/4DxiiBjixHZVjc69WpJX95/4708b0bdc8e713faf69a667f8266d190/472182",
      "name": "Contentful",
      "url": "https://www.contentful.com",
    },
    "categories": Array [
      "Featured",
      "Deployment & Delivery",
    ],
    "description": "### About the app

The Netlify app makes creating and updating content on your site easier. The app removes the need to navigate between platforms, giving editors full control of everything from previewing to deploying changes. Content writers and editors can trigger Netlify builds with one click in the sidebar. 

See build status in real-time, track progress and view recent changes all from within the web app. 


---

### What you can do with the Netlify app: 

- Trigger Netlify builds with one click in the sidebar
- Display build status so editors don't preview outdated sites
- Bring Netlify functionality into the Contentful web app
- Configure on a per-content-type level 

### Requirements:

- A static site based on a static site generator such as Gatsby, Middleman or Hugo
- Your site configured to load content from Contentful and its code pushed to a remote repository, such as GitHub or Gitlab
- A Netlify setup where the site is connected to the remote repository and enabled for continuous deployment

---

### Screenshots

![Netlify config](//images.ctfassets.net/lpjm8d10rkpy/4Rf8yoGPw4AQwdMboiVwhF/a0a9a8b8e73a5ace1fdf78adebc24a49/apps-netlify-site-setup.b4a24a69c0.png)

![Netlify sidebar](//images.ctfassets.net/lpjm8d10rkpy/RrSqXTcdlpfkPt7sJzkEb/57172d0fd3eaa36a5ab90e7d183ed171/apps-netlify-sidebar.0ffd73e12b.png) ",
    "icon": "//images.ctfassets.net/lpjm8d10rkpy/2dYQ3x6VMNZJgDXYmMzLBx/b338ad4b50cd0ff31e7157671b801b4d/logomark.svg",
    "id": "netlify",
    "isEarlyAccess": false,
    "isListed": true,
    "legal": Object {
      "eula": "https://github.com/contentful/apps/blob/master/apps/netlify/LICENSE",
      "privacyPolicy": "https://www.contentful.com/legal/us/privacy/",
    },
    "links": Array [
      Object {
        "title": "Netlify Docs",
        "url": "https://www.netlify.com/docs/",
      },
      Object {
        "title": "Documentation for the Netlify app",
        "url": "https://www.contentful.com/developers/docs/extensibility/apps/netlify/",
      },
      Object {
        "title": "Netlify app source code on GitHub",
        "url": "https://github.com/contentful/apps/tree/master/apps/netlify",
      },
    ],
    "tagLine": "Easily build and preview sites with the Netlify app ",
    "title": "Netlify",
  },
  Object {
    "appDefinition": Object {
      "locations": Array [
        "app",
        "entry-field",
      ],
      "name": "My app",
      "src": "http://localhost:666",
      "sys": Object {
        "id": "app-definition-id",
        "type": "AppDefinition",
      },
    },
    "appInstallation": undefined,
    "id": "app-definition-id",
    "isEarlyAccess": false,
    "isPrivateApp": true,
    "title": "My app",
  },
]
`;
